### PR TITLE
[rollout, diffusion] feat: support NFT and DPO step execution

### PR DESCRIPTION
**Pull request by Sky-Trigger**
### What does this PR do?

This PR extends Qwen-Image continuous batching (`step_execution=True`) to **DiffusionNFT** and **online DPO**, building on the existing FlowGRPO and MixGRPO integrations.

It:

- Adds request-local step-execution state preparation for DiffusionNFT and DPO.
- Preserves each algorithm's existing training-output contract:
  - DiffusionNFT: `latents_clean`, `train_timesteps`, and prompt embeddings/masks.
  - Online DPO: `latents_clean` and prompt embeddings/masks.
- Keeps DPO live latents homogeneous in float32 during continuous batching.
- Reuses the standard Qwen-Image denoising and scheduler path for DiffusionNFT.
- Supports tokenized rollout prompts and raw-text warm-up requests.
- Adds real-engine output-contract tests for both algorithms.
- Updates the step-execution integration guide with algorithm-specific state, precision, and output requirements.

No new algorithm registration or `_stepwise` variant is introduced. The existing `diffusion_nft` and `dpo` algorithm names are used.

### Checklist Before Starting

- [x] Search for similar PRs. Query: [Qwen-Image step execution PRs](https://github.com/verl-project/verl-omni/pulls?q=is%3Apr+Qwen-Image+%22step+execution%22)
- [x] Format the PR title as `[{modules}] {type}: {description}`.

#### Automated coverage

Added real-engine step-execution contract tests for DiffusionNFT and online DPO:

```bash
pytest tests/workers/rollout/rollout_vllm/test_vllm_omni_generate.py \
  -k "diffusion_nft_step_execution_contract or dpo_step_execution_contract"
```

The tests validate:

- Successful generation with `step_execution=True`.
- Presence and shape of algorithm-specific training tensors.
- Float32 `latents_clean`.
- Prompt embedding/mask consistency.
- DiffusionNFT `train_timesteps`.
- Preservation of optional negative-prompt fields.

#### DiffusionNFT training and rollout timing

The following curves show the DiffusionNFT reward and rollout time with step execution enabled.

af6638f254e3d803f46749a5fd9bf900
7579c7f2dcbe9e7861f6fcc3541f8ac0

Experiment configuration: `<model / GPU count / batch size / training steps / rollout settings>`

Observation: `<briefly summarize reward behavior, training stability, and rollout time>`

#### Online DPO training and rollout timing

The following curves show the online DPO reward and rollout time with step execution enabled.

686bf5bad5ff4a1a5aa8d894aed03ee6
4bd45675a53ede374854aa769ad7ebc1

Experiment configuration: `<model / GPU count / batch size / training steps / rollout settings>`

Observation: `<briefly summarize reward behavior, training stability, and rollout time>`

### API and Usage Example

No new API is introduced. Enable the existing step-execution mode and select the existing algorithm registration:

```yaml
actor_rollout_ref:
  model:
    algorithm: diffusion_nft  # or: dpo

  rollout:
    step_execution: true
    max_num_seqs: 16
```

#### DiffusionNFT

- Uses a shared generation-context helper for both step and non-step paths.
- Creates request-local scheduler state for continuous batching.
- Reuses the standard Qwen-Image `denoise_step()` and `step_scheduler()` implementations.
- Reconstructs the existing DiffusionNFT training payload during `post_decode()`:
  - `latents_clean`
  - `train_timesteps`
  - prompt embeddings and masks

#### Online DPO

- Adds algorithm-specific `prepare_encode()`, `denoise_step()`, `step_scheduler()`, and `post_decode()` hooks.
- Generates initial noise using the prompt-embedding dtype before converting live step state to float32.
- Casts latents to the transformer compute dtype only for model execution.
- Keeps scheduler inputs and live request latents in float32 to avoid mixed-dtype batches across concurrently scheduled requests.
- Returns the same final-latent training payload consumed by the non-step DPO path.

#### Tests and documentation

- Extends the real vLLM-Omni rollout test server to select `diffusion_nft` or `dpo`.
- Adds output-contract tests for both algorithms.
- Documents algorithm-specific precision, state, and training-output requirements.

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md).
- [x] Run `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`.
- [x] Update the documentation.
- [x] Add real-engine tests covering DiffusionNFT and online DPO step execution.

Co-authored-by: GitHub Copilot

### CODE REVIEW

**Comment by SamitHuang** [comment]
@AndyZhou952  PTAL

**Review by zhtmike — APPROVED** [review]
LGTM

File: docs/contributing/integrating_a_stepwise_continuous_batching_model.md [hunk 1]
```diff
@@ -211,23 +216,38 @@ the full-forward path.
 
 ### Latents and timesteps
 
-Initial latents should be created in float32:
+Choose the live-latent dtype from the algorithm's scheduler and full-forward
+semantics, and keep it homogeneous across all in-flight requests.
+
+FlowGRPO, MixGRPO, and online DPO preserve seeded parity with their full-forward
```
Comment (wtomin) [inline docs/contributing/integrating_a_stepwise_continuous_batching_model.md]: In `verl_omni/pipelines/qwen_image_flow_grpo/vllm_omni_rollout_adapter.py`, I find that:
```python
  latents = self.prepare_latents(
      ...
      torch.float32,
      ...
  )
```
It is not aligned with the statement here.

File: verl_omni/pipelines/qwen_image_dpo/vllm_omni_rollout_adapter.py [hunk 2]
```diff
@@ -40,6 +43,210 @@ def __init__(self, *, od_config: OmniDiffusionConfig, prefix: str = ""):
...
+
+    def post_decode(
+        self,
+        state: DiffusionRequestState,
+        **kwargs: Any,
+    ) -> DiffusionOutput:
+        """Decode step execution output with DPO training tensors."""
+        del kwargs
+        self._current_timestep = None
+        output = self._decode_latents(state.latents, state.extra["height"], state.extra["width"], "pil")
```
Comment (wtomin) [inline verl_omni/pipelines/qwen_image_dpo/vllm_omni_rollout_adapter.py:237]: Why not getting the `output_type` from kwargs? As you did in `qwen_image_diffusion_nft`

```output = self._decode_latents(state.latents, height, width, kwargs.get("output_type") or "pil")```

File: verl_omni/pipelines/qwen_image_dpo/vllm_omni_rollout_adapter.py [hunk 3]
```diff
@@ -40,6 +43,210 @@ def __init__(self, *, od_config: OmniDiffusionConfig, prefix: str = ""):
...
+        num_inference_steps = sampling.num_inference_steps or 50
+        num_images_per_prompt = sampling.num_outputs_per_prompt if sampling.num_outputs_per_prompt > 0 else 1
+        true_cfg_scale = _coalesce_not_none(sampling.true_cfg_scale, 4.0)
+        max_sequence_length = sampling.max_sequence_length or 512
+
+        generator = sampling.generator
+        if generator is None and sampling.seed is not None:
+            generator = torch.Generator(device=self.device).manual_seed(sampling.seed)
+
+        self._guidance_scale = 1.0
```
Comment (wtomin) [inline verl_omni/pipelines/qwen_image_dpo/vllm_omni_rollout_adapter.py:107]: Why not use `guidance_scale = sampling.guidance_scale if sampling.guidance_scale_provided else 1.0`?

**Comment by wtomin** [comment]
@Sky-Trigger I gave some minor comments. The rest looks fine to me.

File: verl_omni/pipelines/qwen_image_dpo/vllm_omni_rollout_adapter.py (hunk 2 above)
Comment (Sky-Trigger) [reply verl_omni/pipelines/qwen_image_dpo/vllm_omni_rollout_adapter.py:237]: This simply aligns with the existing non-step path. If this behavior needs to be corrected, we can update both the step and non-step paths together in a follow-up.

File: verl_omni/pipelines/qwen_image_dpo/vllm_omni_rollout_adapter.py (hunk 3 above)
Comment (Sky-Trigger) [reply verl_omni/pipelines/qwen_image_dpo/vllm_omni_rollout_adapter.py:107]: Same as above.

File: docs/contributing/integrating_a_stepwise_continuous_batching_model.md (hunk 1 above)
Comment (Sky-Trigger) [reply docs/contributing/integrating_a_stepwise_continuous_batching_model.md]: I have updated this part.

**Review by SamitHuang — APPROVED** [review]
